shared_model/interfaces/common_objects/impl/amount.cpp

```diff
@@ -5,49 +5,73 @@
 
 #include "interfaces/common_objects/amount.hpp"
 
-#include <regex>
-
+#include <boost/algorithm/string/classification.hpp>
 #include <boost/multiprecision/cpp_int.hpp>
 #include "utils/string_builder.hpp"
 
 static const char kDecimalSeparator = '.';
 static const char *kDigits = "0123456789";
+static const char kZero = kDigits[0];
 
 using namespace shared_model::interface;
 
 struct Amount::Impl {
   Impl(const std::string &amount)
       : string_repr_("NaN"), precision_(0), multiprecision_repr_(0) {
-    const auto dot_pos = amount.find_first_not_of(kDigits);
-    if (dot_pos != std::string::npos) {
-      // fail, if:
-      if (amount[dot_pos]
-              != kDecimalSeparator  // string contains an invalid character
-          or amount.find_first_not_of(kDigits, dot_pos + 1)
-              != std::string::npos  // string contains more than one non-digit
-          or dot_pos == 0  // dot is the first symbol (for compatibility)
-          or dot_pos == amount.size() - 1  // dot is the last symbol
-      ) {
+    const char *const start = amount.data();
+    const char *const end = start + amount.size();
+
+    const char *first_nonzero_digit_pos = end;
+    const char *dot_pos = end;
+    for (auto c = start; c < end; ++c) {
```

Comment:
```suggestion
    for (auto *c = start; c < end; ++c) {
```
This would greatly improve readability.